src/main/kotlin/persistence/repo/UserRepo.kt

```diff
@@ -0,0 +1,171 @@
+            it.onNext(
+                    userEntityDao.fetchById(id).first()
+            )
+        }.map {
+            userMapper.mapFromEntity(it)
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * gets all the users currently stored in db
+     */
+    override fun getAll(): Observable<List<User>> {
+        return Observable.create<List<UserEntity>> {
+            it.onNext(
+                    userEntityDao.findAll()
+            )
+        }.map {
+            it.map { userMapper.mapFromEntity(it) }
+        }.subscribeOn(Schedulers.io())
+
+    }
+
+    override fun update(user: User): Completable {
+        return Completable.fromAction {
+            userEntityDao.update(userMapper.mapToEntity(user))
+        }.subscribeOn(Schedulers.io())
+    }
+
+    // considering return a new update user that contains the language for ease of use
+    override fun addLanguage(user: User, language: Language, isSource: Boolean) : Completable {
+        return Completable.fromAction {
+            userLanguageRepo.insert(
+                    UserLanguagesEntity(
+                            user.id,
+                            language.id,
+                            if(isSource) 1 else 0
+                    )
+            ).blockingFirst()
+        }.subscribeOn(Schedulers.io())
+    }
+
+    override fun removeLanguage(user: User, language: Language, isSource: Boolean) : Completable {
+        return Completable.fromAction {
+            userLanguageRepo.delete(
+                    UserLanguagesEntity(
+                            user.id,
+                            language.id,
+                            if(isSource) 1 else 0
+                    )
+            ).blockingAwait()
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * deletes user by id
+     */
+    override fun delete(user: User): Completable {
+        return Completable.fromAction {
+            userEntityDao.delete(userMapper.mapToEntity(user))
+        }.subscribeOn(Schedulers.io())
+    }
+
+    override fun setLanguagePreference(user: User, language: Language, isSource: Boolean) : Completable {
+        return Completable.fromAction {
+            val userPreferences = userPreferencesEntityDao.fetchByUserfk(user.id).first()
+            // updates either the preferred source or target language
+            if (isSource) {
+                userPreferences.sourcelanguagefk = language.id
+            } else {
+                userPreferences.targetlanguagefk = language.id
+            }
+            userPreferencesEntityDao.update(userPreferences)
+        }.subscribeOn(Schedulers.io())
+    }
+
+    /**
+     * This function should be unnecessary because we have add and remove function
+     * this then enforces clients to add and remove languages through the database
+     * keeping just in case we want this
+     */
+    /*
+    private fun updateUserLanguageReferences(user: User, userId: Int) {
```

Comment:
I left this is in just in case we allow the front end to update language references by creating a new user with a list of new references, however I think its not what we want and is an unnecessary function to have